### PR TITLE
T6961: package-build: Fix kea build race condition

### PR DESCRIPTION
**Pull request by MattKobayashi**
## Change Summary

Fix a race condition in the `kea` package build process.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

https://vyos.dev/T6961

## Component(s) name

package-build (`kea`)

## Proposed changes

Because the `package.toml` file for the `kea` package doesn't specify a build command, it defaults to running `dpkg-buildpackage -uc -us -tc -F` against the cloned `git` repo. This is a problem because, by default, `dpkg-buildpackage` runs tasks in parallel (the limit is defined as `auto`, but seems to be based off the number of processors the system has), and there's a component being compiled in `kea` that other make tasks rely on, but ends up sometimes not being built in time for the dependent task to run successfully.

This PR adds a build command to kea's package.toml that specifies the additional argument `-j 1` to the original `dpkg-buildpackage` command, which reverts `dpkg-buildpackage`'s behaviour to serial execution of tasks.

## How to test

```
(using the modified `package.toml`)
git clone https://github.com/vyos/vyos-build
cd vyos-build/scripts/package-build/kea
./build.py
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-build/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sever-sever** [comment]
The logic is
1. it tries to check if build_cmd exists; if not command, it tries to build source + binary
2. If sources fail, it tries to build binaries only

https://github.com/vyos/vyos-build/blob/c1f9fc6a6d387648e91a74a72b3488c42a6b897a/scripts/package-build/build.py#L117C1-L126C65

The similar logic as for Jenkins https://github.com/vyos/vyos-build/blob/c1f9fc6a6d387648e91a74a72b3488c42a6b897a/vars/cloneAndBuild.groovy#L67-L71

File: scripts/package-build/kea/package.toml [hunk 1]
```diff
@@ -2,3 +2,4 @@
 name = "isc-kea"
 commit_id = "debian/2.4.1-3"
 scm_url = "https://salsa.debian.org/debian/isc-kea"
+build_cmd = "dpkg-buildpackage -uc -us -tc -F -j 1"
```
Comment (sever-sever) [inline scripts/package-build/kea/package.toml]: Build fails
Did you check it in `vyos-build` container?
```
(Reading database ... 67171 files and directories currently installed.)
Preparing to unpack isc-kea-build-deps_2.4.1-3_all.deb ...
Unpacking isc-kea-build-deps (2.4.1-3) over (2.4.1-3) ...
Setting up isc-kea-build-deps (2.4.1-3) ...
dpkg-buildpackage: error: unknown option or argument 1

Use --help for program usage information.
Command 'dpkg-buildpackage -uc -us -tc -F -j 1' returned non-zero exit status 2.
I: Source packages build failed, ignoring - building binaries only
dpkg-buildpackage: error: unknown option or argument 1

Use --help for program usage information.
Failed to build package isc-kea: Command 'dpkg-buildpackage -uc -us -tc -F -j 1' returned non-zero exit status 2.
I: Cleaned up build dependency packages
vyos_bld@ce1d3ca789eb:/vyos/work/tmp/tmp-aws/vyos-build/scripts/package-build/kea$ 

```
With this change:
```
vyos_bld@ce1d3ca789eb:/vyos/work/tmp/tmp-aws/vyos-build/scripts/package-build/kea$ git diff .
diff --git a/scripts/package-build/kea/package.toml b/scripts/package-build/kea/package.toml
index 0bfce21e..1b06714b 100644
--- a/scripts/package-build/kea/package.toml
+++ b/scripts/package-build/kea/package.toml
@@ -2,3 +2,4 @@
 name = "isc-kea"
 commit_id = "debian/2.4.1-3"
 scm_url = "https://salsa.debian.org/debian/isc-kea"
+build_cmd = "dpkg-buildpackage -uc -us -tc -F -j 1"



```

Comment (MattKobayashi) [reply scripts/package-build/kea/package.toml]: Minor typo, just needs the space removed between the flag and the value. Will push a fix now.

Comment (sever-sever) [reply scripts/package-build/kea/package.toml]: The question is the same.
Could you let me know if you checked it in the container vyos-build? 
It still fails : 
```
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build .
dpkg-source: info: using options from isc-kea/debian/source/local-options: --unapply-patches
dpkg-source: info: using patch list from debian/patches/series
dpkg-source: info: applying 0002-kea_admin_fix.patch
dpkg-source: info: applying 0009-disable-database-tests.patch
dpkg-source: info: applying 0010-set-control-sockets-location.patch
dpkg-source: info: applying 0011-kea-ctrl-agent-authentication.patch
 debian/rules clean
dh clean --with python3 -X.la
   dh_auto_clean -O-X.la
   dh_clean -O-X.la
 dpkg-source -b .
dpkg-source: info: using options from isc-kea/debian/source/local-options: --unapply-patches
dpkg-source: error: can't build with source format '3.0 (quilt)': no upstream tarball found at ../isc-kea_2.4.1.orig.tar.{bz2,gz,lzma,xz}
dpkg-buildpackage: error: dpkg-source -b . subprocess returned exit status 255
Command 'dpkg-buildpackage -uc -us -tc -F -j1' returned non-zero exit status 2.
I: Source packages build failed, ignoring - building binaries only
dpkg-buildpackage: info: source package isc-kea
dpkg-buildpackage: info: source version 2.4.1-3
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Paride Legovini <paride@debian.org>
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build .
dpkg-source: info: using options from isc-kea/debian/source/local-options: --unapply-patches
 debian/rules clean
dh clean --with python3 -X.la
   dh_auto_clean -O-X.la
   dh_clean -O-X.la
 dpkg-source -b .
dpkg-source: info: using options from isc-kea/debian/source/local-options: --unapply-patches
dpkg-source: error: can't build with source format '3.0 (quilt)': no upstream tarball found at ../isc-kea_2.4.1.orig.tar.{bz2,gz,lzma,xz}
dpkg-buildpackage: error: dpkg-source -b . subprocess returned exit status 255
Failed to build package isc-kea: Command 'dpkg-buildpackage -uc -us -tc -F -j1' returned non-zero exit status 2.
I: Cleaned up build dependency packages
vyos_bld@ca4be1dd41f2:/vyos/work/tmp/vyos-build/scripts/package-build/kea$ 

```

Comment (MattKobayashi) [reply scripts/package-build/kea/package.toml]: There were a number of issues, but I've finally got it building in the container. Please check.

Comment (sever-sever) [reply scripts/package-build/kea/package.toml]: build in container without any changes work fine 
What exactly are you trying to fix?

Comment (MattKobayashi) [reply scripts/package-build/kea/package.toml]: The issue is thoroughly explained in the Phabricator task, I'm not quite sure how to explain it any more clearly. Builds don't always fail as-is, but they do fail sometimes because of the race condition. There is also a warning when building Kea to avoid using parallelism when compiling.

**Comment by c-po** [comment]
This branch has conflicts that must be resolved

File: scripts/package-build/kea/package.toml (hunk 1 above)
Comment (sever-sever) [reply scripts/package-build/kea/package.toml]: Sorry, I cannot reproduce it. It tried to build sources+binaries => fails => builds binaries without any issue.

It clones the repo, checkouts the commit ID, and then builds and installs the dependency and patches to the repo after it builds binaries.
All logic works fine without any change.
```
vyos_bld@cba24cf57398:/vyos/work/tmp/vyos-build/scripts/package-build/kea$ ./build.py 
Cloning into 'isc-kea'...
warning: redirecting to https://salsa.debian.org/debian/isc-kea.git/
remote: Enumerating objects: 13609, done.
remote: Counting objects: 100% (7894/7894), done.
remote: Compressing objects: 100% (2580/2580), done.
remote: Total 13609 (delta 5284), reused 7866 (delta 5266), pack-reused 5715 (from 1)
Receiving objects: 100% (13609/13609), 22.25 MiB | 5.82 MiB/s, done.
Resolving deltas: 100% (8492/8492), done.
Note: switching to 'debian/2.4.1-3'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at cb1f1e5 Update changelog for 2.4.1-3 release
I: Patch directory patches/isc-kea does not exist, skipping patch application
I: Tarball created: isc-kea_debian_2.4.1-3.tar.gz
dpkg-buildpackage: info: source package isc-kea-build-deps
dpkg-buildpackage: info: source version 2.4.1-3
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Equivs Dummy Package Generator <root@cba24cf57398>
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build .
 debian/rules clean
dh clean
   dh_clean
 debian/rules binary
dh binary
   dh_update_autotools_config
   dh_autoreconf
   create-stamp debian/debhelper-build-stamp
   dh_prep
   dh_auto_install --destdir=debian/isc-kea-build-deps/
   dh_install
   dh_installdocs
   dh_installchangelogs
   dh_perl
   dh_link
   dh_strip_nondeterminism
   dh_compress
   dh_fixperms
   dh_missing
   dh_installdeb
   dh_gencontrol
   dh_md5sums
   dh_builddeb
dpkg-deb: building package 'isc-kea-build-deps' in '../isc-kea-build-deps_2.4.1-3_all.deb'.
 dpkg-genbuildinfo --build=binary -O../isc-kea-build-deps_2.4.1-3_amd64.buildinfo
 dpkg-genchanges --build=binary -O../isc-kea-build-deps_2.4.1-3_amd64.changes
dpkg-genchanges: info: binary-only upload (no source code included)
 dpkg-source --after-build .
dpkg-buildpackage: info: binary-only upload (no source included)

...
```
And at the end, we have binaries only:
```
vyos_bld@cba24cf57398:/vyos/work/tmp/vyos-build/scripts/package-build/kea$ ls -lah | grep \.deb
-rw-r--r--  1 vyos_bld vyos_bld  33M Dec 23 09:00 isc-kea_debian_2.4.1-3.tar.gz
-rw-r--r--  1 vyos_bld vyos_bld 214K Dec 23 09:10 kea_2.4.1-3_all.deb
-rw-r--r--  1 vyos_bld vyos_bld 434K Dec 23 09:10 kea-admin_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 4.1M Dec 23 09:10 kea-admin-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 3.9M Dec 23 09:10 kea-common_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld  72M Dec 23 09:10 kea-common-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 347K Dec 23 09:10 kea-ctrl-agent_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 2.4M Dec 23 09:10 kea-ctrl-agent-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 823K Dec 23 09:10 kea-dev_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 211K Dec 23 09:10 kea-dev-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 557K Dec 23 09:10 kea-dhcp4-server_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 6.4M Dec 23 09:10 kea-dhcp4-server-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 554K Dec 23 09:10 kea-dhcp6-server_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 6.4M Dec 23 09:10 kea-dhcp6-server-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 369K Dec 23 09:10 kea-dhcp-ddns-server_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 3.5M Dec 23 09:10 kea-dhcp-ddns-server-dbgsym_2.4.1-3_amd64.deb
-rw-r--r--  1 vyos_bld vyos_bld 3.8M Dec 23 09:10 kea-doc_2.4.1-3_all.deb
-rw-r--r--  1 vyos_bld vyos_bld 216K Dec 23 09:10 python3-kea-connector_2.4.1-3_all.deb
vyos_bld@cba24cf57398:/vyos/work/tmp/vyos-build/scripts/package-build/kea$ 

```
Are there any issues with this step?

Yes you don't build source but build binaries only in this case.
```
dpkg-source: info: using options from isc-kea/debian/source/local-options: --unapply-patches
dpkg-source: error: can't build with source format '3.0 (quilt)': no upstream tarball found at ../isc-kea_2.4.1.orig.tar.{bz2,gz,lzma,xz}
dpkg-buildpackage: error: dpkg-source -b . subprocess returned exit status 255
Command 'dpkg-buildpackage -uc -us -tc -F' returned non-zero exit status 2.
I: Source packages build failed, ignoring - building binaries only

```
I will happy to re-test the PR after rebase/conflict solvind

File: scripts/package-build/build.py [hunk 2]
```diff
@@ -32,30 +32,32 @@ def ensure_dependencies(dependencies: list) -> None:
         return
 
     print("I: Ensure Debian build dependencies are met")
-    run(['sudo', 'apt-get', 'update'], check=True)
-    run(['sudo', 'apt-get', 'install', '-y'] + dependencies, check=True)
+    run(["sudo", "apt-get", "update"], check=True)
```
Comment (sever-sever) [inline scripts/package-build/build.py:35]: Do not replace single-quotes

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by MattKobayashi** [comment]
Can't be bothered trying to contribute to this project any more. 🖕🏻

File: scripts/package-build/build.py [hunk 3]
```diff
@@ -147,34 +172,38 @@ def cleanup_build_deps(repo_dir: Path) -> None:
...
+    arg_parser.add_argument(
+        "--config",
+        default="package.toml",
+        help="Path to the package configuration file",
+    )
+    arg_parser.add_argument(
+        "--patch-dir",
+        default="patches",
+        help="Path to the directory containing patches",
+    )
```
Comment (sever-sever) [inline scripts/package-build/build.py:195]: Please don't change the format in the same commit as the fix.
Changing the format should be a separate PR or commit. This way, it will be clear what exactly was changed by fix and what was reformatted.

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: scripts/package-build/build.py (hunk 3 above)
Comment (MattKobayashi) [reply scripts/package-build/build.py:195]: PR is closed.

File: scripts/package-build/build.py (hunk 2 above)
Comment (MattKobayashi) [reply scripts/package-build/build.py:35]: PR is closed.
